### PR TITLE
2.4 Implemented PXB-2608 (Upgrade Vault API to V2) 

### PR DESCRIPTION
**Pull request by altmannmarcelo**
https://jira.percona.com/browse/PXB-2608

Cherry-picked related vault v2 commits from PS 5.7

Adjusted test cases - Splitted test cases into separate file to explore
different secret_mount_point_version options.

Adjusted transition-key size - Vault v2 generates bigger file names (hashed)
which in case of using generate-transition-key would cause vault to
return an error due to file been to big on disk.

Fixed DBUG_ASSRT to assert

### CODE REVIEW

**Comment by altmannmarcelo** [comment]
https://pxb.cd.percona.com/job/percona-xtrabackup-2.4-test-param/73/

Local vault & keyring tests:
```
VAULT_URL=https://192.168.23.44:8200 VAULT_CACERT=${HOME}/vault.crt VAULT_CA=${HOME}/vault.crt VAULT_TOKEN=s.aKCe6mbqQ6Zx9FHcaDYlqgR9 ./run.sh -t t/innodb_keyring_vault_v1.sh -t t/innodb_keyring_vault_v2.sh -t t/innodb_keyring_vault_auto_v1.sh -t t/innodb_keyring_vault_auto_v2.sh -t t/innodb_encryption_mix_plugins.sh -t t/innodb_keyring_file.sh
Using /work/pxb/ins/2.4/xtrabackup-test/var as test root
Detecting server version...
Running against Percona Server 5.7.34-37-debug-log (XtraDB 5.7.34-37)
Using 'xtrabackup' as xtrabackup binary
Autodetected number of cores: 32
Limiting to 16 to avoid excessive resource consumption
Using 16 parallel workers

==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
innodb_encryption_mix_plugins            w5	[passed]    19
innodb_keyring_vault_v2                  w2	[passed]    184
innodb_keyring_vault_auto_v2             w4	[passed]    184
innodb_keyring_vault_v1                  w1	[passed]    187
innodb_keyring_vault_auto_v1             w3	[passed]    187
innodb_keyring_file                      w6	[passed]    200
==============================================================================
Spent 961 of 207 seconds executing testcases

SUMMARY: 6 run, 6 successful, 0 skipped, 0 failed

See results/ for detailed output
```

**Comment by altmannmarcelo** [comment]
@satya-bodapati I Applied the same patch as 8.0 (multi version magic key) with the except of smart pointer as it's not available on gnu++03)

File: storage/innobase/xtrabackup/src/keyring_plugins.cc [hunk 1]
```diff
@@ -93,7 +94,11 @@ xb_fetch_tablespace_key(ulint space_id, byte *key, byte *iv)
 const char *TRANSITION_KEY_PREFIX = "XBKey";
 const size_t TRANSITION_KEY_PREFIX_LEN = sizeof(TRANSITION_KEY_PREFIX) / sizeof(char);
 const size_t TRANSITION_KEY_RANDOM_DATA_LEN = 32;
-const size_t TRANSITION_KEY_NAME_MAX_LEN = TRANSITION_KEY_PREFIX_LEN + ENCRYPTION_SERVER_UUID_LEN + TRANSITION_KEY_RANDOM_DATA_LEN + 1;
+const size_t TRANSITION_KEY_NAME_MAX_LEN_V1 =
+    ENCRYPTION_SERVER_UUID_LEN + 2 + 45;
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/keyring_plugins.cc:98]: can you comment or improve how this 45 is arrived at?

File: storage/innobase/xtrabackup/src/keyring_plugins.cc [hunk 2]
```diff
@@ -528,17 +540,20 @@ xb_tablespace_keys_load_one(const char *dir, const char *transition_key,
 		goto error;
 	}
 
-	if (fread(transition_key_name, 1, sizeof(transition_key_name), f)
-		!= sizeof(transition_key_name)) {
+	transition_key_name =
+	(char *)my_malloc(PSI_NOT_INSTRUMENTED, transition_key_name_size,
+			MYF(MY_FAE | MY_ZEROFILL));
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/keyring_plugins.cc:545]: I assume c+11 is not supported in 2.4?

**Review by satya-bodapati — APPROVED** [review]

File: storage/innobase/xtrabackup/src/keyring_plugins.cc (hunk 1 above)
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/src/keyring_plugins.cc:98]: @satya-bodapati this is the length of random key at https://github.com/percona/percona-xtrabackup/blob/percona-xtrabackup-2.4.24/storage/innobase/xtrabackup/src/keyring_plugins.cc#L140

File: storage/innobase/xtrabackup/src/keyring_plugins.cc (hunk 2 above)
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/src/keyring_plugins.cc:545]: Correct, for 2.4 we use gnu++03 standard

File: storage/innobase/xtrabackup/src/keyring_plugins.cc (hunk 1 above)
Comment (satya-bodapati) [reply storage/innobase/xtrabackup/src/keyring_plugins.cc:98]: I see. rand64 was not 64 bytes :) Just 45

**Review by satya-bodapati — APPROVED** [review]

**Pull request merged**